py/mattak/backends/uproot/dataset.py

```diff
@@ -203,7 +255,88 @@ def eventInfo(self) -> Union[Optional[mattak.Dataset.EventInfo],Sequence[Optiona
+        (v_residuals, residuals) : tuple of numpy arrays
+            both v_residuals and residuals have shape (points, 2)
+
+        """
+
+        vres_dac1 = cal_file["aveResid_dac1"].values(axis = 0)
+        vres_dac2 = cal_file["aveResid_dac2"].values(axis = 0)
+        residual_dac1 = cal_file["aveResid_dac1"].values(axis = 1)
+        residual_dac2 = cal_file["aveResid_dac2"].values(axis = 1)
+        return numpy.stack(numpy.array([vres_dac1, vres_dac2]), axis = -1), numpy.stack(numpy.array([residual_dac1, residual_dac2]), axis = -1)
+    
+    def __calibrate(self, waveform_array : numpy.ndarray, param : numpy.ndarray, 
+                    vres : numpy.ndarray, res : numpy.ndarray, starting_window : float | int,
+                    fit_min : float = -1.3, fit_max : float = 0.7, accuracy : float = 0.005) -> numpy.ndarray:
+        """
+        The calibration function that transforms waveforms from ADC to voltage
+
+        Parameters
+        ----------
+        waveform_array : array of shape (24, 2048)
+            array of one waveform
+        param : array of shape (24 * 4096, 10)
+            the parameters found in the calibration file
+        vres : array of shape (points, 2)
+            the voltage points of the residuals, shape
+        res : array of shape (points, 2)
+            the ADC values of the residuals
+        starting_window : int | float
+            the sample on which the run started
+        fit_min : float
+            lower bound of original fit used on the bias scan
+        fit_max : float
+            upper bound of original fit used on the bias scan
+        accuracy : float
+            nr of voltage steps in table of the calibration function
+
+        Return
+        ------
+        waveform_volt : array of shape (24, 2048)
+            calibrated waveform in volt
+        """
+
+        # "discrete" inverse
+        vsamples = numpy.arange(fit_min, fit_max, accuracy)
+        waveform_volt = numpy.zeros((24, 2048))
+        # residuals split over DACs    
+        ressamples = (numpy.interp(vsamples, vres[:, 0], res[:, 0]), numpy.interp(vsamples, vres[:, 1], res[:, 1]))
+        
+        for c, wf_channel in enumerate(waveform_array):
+            starting_window_channel = starting_window[c]
+            # Reordering the parameters to match the correct starting window
+            # (calibration is done per sample)
+            param_channel = param[4096 * c : 4096 * (c + 1)]
+            samples_idx = (128 * starting_window_channel + numpy.arange(2048)) % 2048
+            if starting_window_channel >= 16:
+                samples_idx += 2048
+            param_channel = param_channel[samples_idx]
+
+            for s, (adc, p) in enumerate(zip(wf_channel, param_channel)):
+                # discrete inverse
+                adcsamples = numpy.polyval(p[::-1], vsamples) + ressamples[int(c/12)]
+                volt = numpy.interp(adc, adcsamples, vsamples, left = fit_min, right = fit_max)
```

Comment:
The entire code might be faster when you use a comprehensive list? Otherwise (against my previous comment to you) it might not be possible to optimise that by not using a for-loop). Other thought: Do we want to use `left` and `right`? Should we not try to do a linear interpolation there?